Feature/sign in

Ammar-IT99 opened the pull request:
add sign-in screen

Mohamed-Samir-Ibrahim committed:
feat: add sign-in screen, cache_network.dart, signin_request.dart, signin_response.dart,signin_use_case.dart,login_screen.dart,login_screen_view_model.dart and states.dart

fix: di.dart, api_service.dart, main.dart, api_constant.dart, pubspec.yaml, auth_remote_data_source_impl.dart, auth_repository_impl.dart, auth_remote_data_source.dart and auth_repository_contract.dart

Ammar-IT99 committed:
Merge branch 'refs/heads/development' into feature/sign-in

# Conflicts:
#	lib/data/api/api_service.dart
#	pubspec.yaml

Ammar-IT99 committed:
feat: **register_screen.dart**: Navigate to login screen.

john-safwat commented on lib/data/api/api_service.dart:
```diff
@@ -19,6 +23,7 @@ class ApiService {
   }
 
+  //register
   Future<Either<Failures, RegisterResponseDataModel>> register(
```
it's not recommended to use either package according to it's restrictions

Ammar-IT99 replied:
```suggestion
  Future<Result<Failures, RegisterResponseDataModel>> register(
```

john-safwat commented on lib/data/api/api_service.dart:
```diff
@@ -66,6 +71,52 @@ class ApiService {
       return Left(ServerError(errorMessage: 'Request timed out'));
+
+    } catch (e) {
+      print('Unexpected Error: $e');
```
don't use prints in production code

john-safwat commented on lib/data/api/api_service.dart:
```diff
@@ -66,6 +71,52 @@ class ApiService {
+      );
+
+      var requestBody = jsonEncode(signIn.toJson());
+      print('Request body: $requestBody');
+      var response = await http.post(
+        url,
+        body: requestBody,
+        headers: {
+          'Content-Type': 'application/json',
+          'Authorization': 'Bearer ${ApiConstant.token}'
+        },
+      );
+      var responseData = jsonDecode(response.body);
+      await CacheNetwork.insertToCache(key: "token", value: responseData['token']??'');
+        ApiConstant.token = await CacheNetwork.getCacheData(key: "token");
+      print('Response status: ${response.statusCode}');
+      print('Response body: ${response.body}');
+      print('user token: ${ApiConstant.token}');
```
don't use prints in production code

john-safwat commented on lib/core/di.dart:
user service locator like getit or injectable as wrapper on getit insted of manual injection

john-safwat commented on lib/presentation/auth/login/cubit/login_screen_view_model.dart:
```diff
@@ -0,0 +1,36 @@
+
+  var formKey = GlobalKey<FormState>();
+  var emailController = TextEditingController(text: 'ampar133@1elevate.com');
+  var passwordController = TextEditingController(text: 'Elevate@123');
```
don't add initial value duo to security issues when the memory is dumbed just create the controller object

john-safwat commented on lib/presentation/auth/login/login_screen.dart:
```diff
@@ -0,0 +1,185 @@
+  void _loadRememberMe() async {
+    final prefs = await SharedPreferences.getInstance();
+    setState(() {
+      rememberMe = prefs.getBool('rememberMe') ?? false;
```
avoid use hard coded strings in your code

john-safwat commented on lib/presentation/auth/login/login_screen.dart:
```diff
@@ -0,0 +1,185 @@
+    final prefs = await SharedPreferences.getInstance();
+    setState(() {
+      rememberMe = prefs.getBool('rememberMe') ?? false;
+      if (rememberMe) {
+        viewModel.emailController.text = prefs.getString('email') ?? '';
+        viewModel.passwordController.text = prefs.getString('password') ?? '';
+      }
+    });
+  }
+
+  void _saveRememberMe() async {
+    final prefs = await SharedPreferences.getInstance();
+    if (rememberMe) {
+      await prefs.setBool('rememberMe', true);
+      await prefs.setString('email', viewModel.emailController.text);
+      await prefs.setString('password', viewModel.passwordController.text);
+    } else {
+      await prefs.remove('rememberMe');
+      await prefs.remove('email');
+      await prefs.remove('password');
+    }
+  }
```
don't store the user email and password shared preferences for the user security

john-safwat commented on lib/presentation/auth/login/login_screen.dart:
```diff
@@ -0,0 +1,185 @@
+        } else if (state is SignInSuccessState) {
+          Navigator.of(context).pop(); // Dismiss the loading dialog
+          ScaffoldMessenger.of(context)
+              .showSnackBar(SnackBar(content: Text("Success")));
```
avoid using hard coded strings

john-safwat commented on lib/presentation/auth/login/login_screen.dart:
```diff
@@ -0,0 +1,185 @@
+                  label: 'Email',
+                  keyboardType: TextInputType.emailAddress,
+                  controller: viewModel.emailController,
+                  validator: (value) {
+                    if (value == null || value.isEmpty) {
+                      return 'Please enter your email';
+                    }
+                    if (!RegExp(r'^[^@]+@[^@]+\.[^@]+').hasMatch(value)) {
+                      return 'Please enter a valid email address';
+                    }
+                    return null;
+                  },
```
don't use hard coded strings instead use localization

john-safwat commented on lib/presentation/auth/login/login_screen.dart:
```diff
@@ -0,0 +1,185 @@
+                Row(
+                  children: [
+                    Expanded(
+                      child: CustomTextFormField(
+                        label: 'Password',
+                        isPassword: true,
+                        controller: viewModel.passwordController,
+                        validator: (value) {
+                          if (value == null || value.isEmpty) {
+                            return 'Please enter your password';
+                          }
+                          if (!RegExp(
+                                  r'^(?=.*?[A-Z])(?=.*?[a-z])(?=.*?[0-9])(?=.*?[#?!@$%^&*-]).{8,}$')
+                              .hasMatch(value)) {
+                            return 'Password must be at least 11 characters long and include an uppercase letter, a lowercase letter, a number, and a special character';
+                          }
+                          return null;
+                        },
+                      ),
```
don't use hard coded strings instead use localization

john-safwat commented on lib/presentation/auth/login/login_screen.dart:
```diff
@@ -0,0 +1,185 @@
+                    if (viewModel.formKey.currentState!.validate()) {
+                      _saveRememberMe();
+                      viewModel.signIn();
+                      print('email + Password = ${viewModel.emailController.text + viewModel.passwordController.text + rememberMe.toString()}');
```
don't use prints in production code

john-safwat commented on lib/presentation/auth/login/login_screen.dart:
```diff
@@ -0,0 +1,185 @@
+                      print('email + Password = ${viewModel.emailController.text + viewModel.passwordController.text + rememberMe.toString()}');
+                    }
+                  },
+                  child: Text('Signin'),
```
con't use hard coded strings in your code instead use localization

john-safwat commented on lib/presentation/auth/login/login_screen.dart:
```diff
@@ -0,0 +1,185 @@
+                Row(
+                  mainAxisAlignment: MainAxisAlignment.center,
+                  children: [
+                  Text(
+                    'Don\'t have an account? ',
+                    style: TextStyle(
+                      color: Colors.black,
+                    ),
+                  ),
+                  TextButton(
+                    onPressed: () {
+                      Navigator.pushNamed(context, RegisterScreen.routeName);
+                    },
+                    child: Text(
+                      'Sign up',
```
con't use hard coded strings in your code instead use localization

john-safwat commented on lib/presentation/auth/register/register_screen.dart:
```diff
@@ -168,7 +169,7 @@ class _RegisterScreenState extends State<RegisterScreen> {
                 ),
                 TextButton(
                   onPressed: () {
-                    // Navigate to login screen
+                  Navigator.pushNamed(context, LoginScreen.routeName);// Navigate to login screen
                   },
                   child: Text('Already have an account? Log in'),
```
this logic is responsibility of view model and don't use hard coded strings in your code

Ammar-IT99 committed:
Update login_screen_view_model.dart

resolve
1-don't add initial value duo to security issues when the memory is dumbed just create the controller object.
2- use switch for sealed class
